### ISSUE TITLE
No provider of glEGLImageTargetTexture2DOES found with NVIDIA proprietary driver

### ISSUE THREAD

**Issue by duffolonious** [anholt/libepoxy#148]
Trying virglrenderer out (expecially since now everything seems to require OpenGL:

> $ virt-viewer --attach test1
> No provider of glEGLImageTargetTexture2DOES found.  Requires one of:
>     GL extension "GL_OES_EGL_image"
> Aborted (core dumped)

Backtrace:

> Core was generated by `virt-viewer --attach test1'.
> Program terminated with signal SIGABRT, Aborted.
> #0  0x00007fa44b0e869b in raise () from /lib64/libc.so.6
> [Current thread is 1 (Thread 0x7fa44fddf5c0 (LWP 11444))]
> 
> Thread 1 (Thread 0x7fa44fddf5c0 (LWP 11444)):
> #0  0x00007fa44b0e869b in raise () at /lib64/libc.so.6
> #1  0x00007fa44b0ea4a0 in abort () at /lib64/libc.so.6
> #2  0x00007fa443bfec9c in gl_provider_resolver () at /lib64/libepoxy.so.0
> #3  0x00007fa443c04bf1 in gl_single_resolver () at /lib64/libepoxy.so.0
> #4  0x00007fa443c09ada in epoxy_glEGLImageTargetTexture2DOES_global_rewrite_ptr () at /lib64/libepoxy.so.0
> #5  0x00007fa44e74a851 in spice_egl_update_scanout () at /lib64/libspice-client-gtk-3.0.so.5
> #6  0x00007fa44e7443ba in spice_display_widget_gl_scanout () at /lib64/libspice-client-gtk-3.0.so.5
> #7  0x00007fa44e749ac7 in spice_egl_init () at /lib64/libspice-client-gtk-3.0.so.5
> #8  0x00007fa44e74435c in spice_display_widget_gl_scanout () at /lib64/libspice-client-gtk-3.0.so.5
> #9  0x00007fa44bd2c30d in g_closure_invoke () at /lib64/libgobject-2.0.so.0
> #10 0x00007fa44bd3e98e in signal_emit_unlocked_R () at /lib64/libgobject-2.0.so.0
> #11 0x00007fa44bd471a5 in g_signal_emit_valist () at /lib64/libgobject-2.0.so.0
> #12 0x00007fa44bd47b0f in g_signal_emit () at /lib64/libgobject-2.0.so.0
> #13 0x00007fa44e1ae726 in gtk_widget_realize () at /lib64/libgtk-3.so.0
> #14 0x00007fa44e1aea28 in gtk_widget_map () at /lib64/libgtk-3.so.0
> #15 0x00007fa44e0fc290 in gtk_stack_forall () at /lib64/libgtk-3.so.0
> #16 0x00007fa44df8b5af in gtk_container_map () at /lib64/libgtk-3.so.0
> #17 0x00007fa44e0fd3e8 in gtk_stack_map () at /lib64/libgtk-3.so.0
> #18 0x00007fa44bd2c450 in _g_closure_invoke_va () at /lib64/libgobject-2.0.so.0
> #19 0x00007fa44bd47494 in g_signal_emit_valist () at /lib64/libgobject-2.0.so.0
> #20 0x00007fa44bd47b0f in g_signal_emit () at /lib64/libgobject-2.0.so.0
> #21 0x00007fa44e1ae9d2 in gtk_widget_map () at /lib64/libgtk-3.so.0
> #22 0x00007fa44df8b5af in gtk_container_map () at /lib64/libgtk-3.so.0
> #23 0x00007fa44dfd947c in gtk_event_box_map () at /lib64/libgtk-3.so.0
> #24 0x00007fa44bd2c450 in _g_closure_invoke_va () at /lib64/libgobject-2.0.so.0
> #25 0x00007fa44bd47494 in g_signal_emit_valist () at /lib64/libgobject-2.0.so.0
> #26 0x00007fa44bd47b0f in g_signal_emit () at /lib64/libgobject-2.0.so.0
> #27 0x00007fa44e1ae9d2 in gtk_widget_map () at /lib64/libgtk-3.so.0
> #28 0x00007fa44df8b5af in gtk_container_map () at /lib64/libgtk-3.so.0
> #29 0x00007fa44bd2c546 in _g_closure_invoke_va () at /lib64/libgobject-2.0.so.0
> #30 0x00007fa44bd47494 in g_signal_emit_valist () at /lib64/libgobject-2.0.so.0
> #31 0x00007fa44bd47b0f in g_signal_emit () at /lib64/libgobject-2.0.so.0
> #32 0x00007fa44e1ae9d2 in gtk_widget_map () at /lib64/libgtk-3.so.0
> #33 0x00007fa44e1aeb88 in gtk_widget_set_child_visible () at /lib64/libgtk-3.so.0
> #34 0x00007fa44e081b7b in gtk_notebook_real_switch_page () at /lib64/libgtk-3.so.0
> #35 0x00007fa44bd2c30d in g_closure_invoke () at /lib64/libgobject-2.0.so.0
> #36 0x00007fa44bd3e3e0 in signal_emit_unlocked_R () at /lib64/libgobject-2.0.so.0
> #37 0x00007fa44bd471a5 in g_signal_emit_valist () at /lib64/libgobject-2.0.so.0
> #38 0x00007fa44bd47b0f in g_signal_emit () at /lib64/libgobject-2.0.so.0
> #39 0x00007fa44e088586 in gtk_notebook_set_current_page () at /lib64/libgtk-3.so.0
> #40 0x000000bfd0cf9745 in virt_viewer_notebook_show_display ()
> #41 0x000000bfd0cf5cd6 in display_show_hint ()
> #42 0x00007fa44bd2c30d in g_closure_invoke () at /lib64/libgobject-2.0.so.0
> #43 0x00007fa44bd3e98e in signal_emit_unlocked_R () at /lib64/libgobject-2.0.so.0
> #44 0x00007fa44bd471a5 in g_signal_emit_valist () at /lib64/libgobject-2.0.so.0
> #45 0x00007fa44bd47b0f in g_signal_emit () at /lib64/libgobject-2.0.so.0
> #46 0x00007fa44bd30594 in g_object_dispatch_properties_changed () at /lib64/libgobject-2.0.so.0
> #47 0x00007fa44bd32a19 in g_object_notify () at /lib64/libgobject-2.0.so.0
> #48 0x000000bfd0d01f99 in update_display_ready ()
> #49 0x00007fa44bd2c30d in g_closure_invoke () at /lib64/libgobject-2.0.so.0
> #50 0x00007fa44bd3e98e in signal_emit_unlocked_R () at /lib64/libgobject-2.0.so.0
> #51 0x00007fa44bd471a5 in g_signal_emit_valist () at /lib64/libgobject-2.0.so.0
> #52 0x00007fa44bd47b0f in g_signal_emit () at /lib64/libgobject-2.0.so.0
> #53 0x00007fa44bd30594 in g_object_dispatch_properties_changed () at /lib64/libgobject-2.0.so.0
> #54 0x00007fa44bd32a19 in g_object_notify () at /lib64/libgobject-2.0.so.0
> #55 0x00007fa44e743d8e in update_area () at /lib64/libspice-client-gtk-3.0.so.5
> #56 0x00007fa44e744193 in spice_display_widget_update_monitor_area () at /lib64/libspice-client-gtk-3.0.so.5
> #57 0x00007fa44bd2c30d in g_closure_invoke () at /lib64/libgobject-2.0.so.0
> #58 0x00007fa44bd3e7ed in signal_emit_unlocked_R () at /lib64/libgobject-2.0.so.0
> #59 0x00007fa44bd471a5 in g_signal_emit_valist () at /lib64/libgobject-2.0.so.0
> #60 0x00007fa44bd47b0f in g_signal_emit () at /lib64/libgobject-2.0.so.0
> #61 0x00007fa44bd30594 in g_object_dispatch_properties_changed () at /lib64/libgobject-2.0.so.0
> #62 0x00007fa44bd32a19 in g_object_notify () at /lib64/libgobject-2.0.so.0
> #63 0x00007fa44ca54030 in notify_main_context () at /lib64/libspice-client-glib-2.0.so.8
> #64 0x00007fa44ba50c27 in g_idle_dispatch () at /lib64/libglib-2.0.so.0
> #65 0x00007fa44ba54247 in g_main_context_dispatch () at /lib64/libglib-2.0.so.0
> #66 0x00007fa44ba545e8 in g_main_context_iterate.isra () at /lib64/libglib-2.0.so.0
> #67 0x00007fa44ba5467c in g_main_context_iteration () at /lib64/libglib-2.0.so.0
> #68 0x00007fa44c00debd in g_application_run () at /lib64/libgio-2.0.so.0
> #69 0x000000bfd0cedc5c in main ()
> 
> Thread 2 (Thread 0x7fa42f638700 (LWP 11445)):
> #0  0x00007fa44b1b6acd in poll () at /lib64/libc.so.6
> #1  0x00007fa44ba54569 in g_main_context_iterate.isra () at /lib64/libglib-2.0.so.0
> #2  0x00007fa44ba5467c in g_main_context_iteration () at /lib64/libglib-2.0.so.0
> #3  0x00007fa44ba546c1 in glib_worker_main () at /lib64/libglib-2.0.so.0
> #4  0x00007fa44ba7b536 in g_thread_proxy () at /lib64/libglib-2.0.so.0
> #5  0x00007fa44b48a36d in start_thread () at /lib64/libpthread.so.0
> #6  0x00007fa44b1c2bbf in clone () at /lib64/libc.so.6
> 
> Thread 3 (Thread 0x7fa427fff700 (LWP 11449)):
> #0  0x00007fa44b1b6acd in poll () at /lib64/libc.so.6
> #1  0x00007fa4444ce831 in linux_udev_event_thread_main () at /lib64/libusb-1.0.so.0
> #2  0x00007fa44b48a36d in start_thread () at /lib64/libpthread.so.0
> #3  0x00007fa44b1c2bbf in clone () at /lib64/libc.so.6
> 
> Thread 4 (Thread 0x7fa42ee37700 (LWP 11446)):
> #0  0x00007fa44b1b6acd in poll () at /lib64/libc.so.6
> #1  0x00007fa44ba54569 in g_main_context_iterate.isra () at /lib64/libglib-2.0.so.0
> #2  0x00007fa44ba54902 in g_main_loop_run () at /lib64/libglib-2.0.so.0
> #3  0x00007fa44c039cb6 in gdbus_shared_thread_func () at /lib64/libgio-2.0.so.0
> #4  0x00007fa44ba7b536 in g_thread_proxy () at /lib64/libglib-2.0.so.0
> #5  0x00007fa44b48a36d in start_thread () at /lib64/libpthread.so.0
> #6  0x00007fa44b1c2bbf in clone () at /lib64/libc.so.6
> 
> Thread 5 (Thread 0x7fa4277fe700 (LWP 11450)):
> #0  0x00007fa44b1b6acd in poll () at /lib64/libc.so.6
> #1  0x00007fa4444c81fc in handle_events () at /lib64/libusb-1.0.so.0
> #2  0x00007fa4444c927a in libusb_handle_events_timeout_completed () at /lib64/libusb-1.0.so.0
> #3  0x00007fa4444c936f in libusb_handle_events () at /lib64/libusb-1.0.so.0
> #4  0x00007fa44ca6a4f1 in spice_usb_device_manager_usb_ev_thread () at /lib64/libspice-client-glib-2.0.so.8
> #5  0x00007fa44ba7b536 in g_thread_proxy () at /lib64/libglib-2.0.so.0
> #6  0x00007fa44b48a36d in start_thread () at /lib64/libpthread.so.0
> #7  0x00007fa44b1c2bbf in clone () at /lib64/libc.so.6

System info (VM host - Fedora 26):

> $ nvidia-smi -q | grep 'Driver Version'
> Driver Version                      : 384.59
> $ rpm -qa libepoxy virglrenderer qemu-system-x86 spice-gtk spice-server mesa-libEGL
> virglrenderer-0.6.0-1.20170210git76b3da97b.fc26.x86_64
> qemu-system-x86-2.9.0-5.fc26.x86_64
> spice-server-0.13.3-2.fc26.x86_64
> spice-gtk-0.34-1.fc26.x86_64
> libepoxy-1.4.3-1.fc26.x86_64
> mesa-libEGL-17.1.7-1.fc26.i686
> mesa-libEGL-17.1.7-1.fc26.x86_64
> 
I followed the steps in the bug reports to get around some other issues (like the DRI device perms):
> sudo chmod 777 /dev/dri/renderD128

Same issue:
https://bugzilla.redhat.com/show_bug.cgi?id=1265997
https://bugzilla.redhat.com/show_bug.cgi?id=1507290

**Comment by ebassi** [comment]
Epoxy is the shim layer providing the trampolines into OpenGL/OpenGL ES. Some functions, in GL, are gated on the existence of an extension; if Epoxy does not find the function, it will `abort()`, to easily let know application developers what went wrong.

It seems virt-viewer is calling a GL function without checking if that function is actually available; virt-viewer should perform a check and either by request a GL implementation that has the extension needed for that function; or it should gracefully degrade its functionality.

This is not an Epoxy issue, as Epoxy is working exactly as intended.

**Comment by duffolonious** [comment]
Ahhh, good to know.  Time to look at virt-viewer then.

**Comment by sjuxax** [comment]
I'm hitting this issue too. It appears that nvidia is not exporting this symbol via conventional means. `es2_info` reports the extension's availability, but `glxinfo` does not. On an Intel GPU with Mesa, I can confirm that this extension is reported by both `glxinfo` and `es2_info`.

Unsure whether this would be considered within the scope of libepoxy, libglvnd, or something else. For now I've just gone back to VFIO passthrough.

The problematic call appears to occur in qemu, not virt-viewer. See https://github.com/qemu/qemu/blob/38d1b31e0501f938db39c5b2e508328530410246/ui/egl-helpers.c#L303.

Output with Mesa:
```
jeff@jeff-yoga:~$ es2_info | grep -i egl_image
    GL_OES_vertex_half_float, GL_EXT_texture_sRGB_decode, GL_OES_EGL_image, 
    GL_OES_EGL_image_external, GL_OES_EGL_sync, GL_OES_vertex_array_object, 
jeff@jeff-yoga:~$ glxinfo | grep -i egl_image
    GL_NV_packed_depth_stencil, GL_NV_texture_barrier, GL_OES_EGL_image, 
    GL_NV_texture_env_combine4, GL_NV_texture_rectangle, GL_OES_EGL_image, 
    GL_OES_EGL_image, GL_OES_EGL_image_external, GL_OES_EGL_sync,
```

Don't have nvidia output right now.

**Comment by duffolonious** [comment]
WIth an NVIDIA card (with a very recent driver too):
> $ nvidia-smi -q | grep -e 'Driver Version' -e 'Product Name'
> Driver Version                      : 384.98
>     Product Name                    : Quadro K2200
> $ es2_info | grep -i egl_image
>     GL_OES_primitive_bounding_box, GL_OES_EGL_image, 
>     GL_OES_EGL_image_external, GL_OES_EGL_image_external_essl3, 
> $ glxinfo | grep -i egl_image

So is this a matter of looking in the wrong lib?  EGL vs GL[X] or something?

**Comment by duffolonious** [comment]
@ebassi -^ is there a way for libepoxy to "failback"?  Is NVIDIA being strict in some sense?
Here is my attempt to make this nice and neat:

> $ diff -Nur <(ldd `which glxinfo` | sort | awk '{$NF=""; print $0}') <(ldd `which es2_info` | sort | awk '{$NF=""; print $0}')
> --- /dev/fd/63	2018-01-02 14:24:19.385693556 -0600
> +++ /dev/fd/62	2018-01-02 14:24:19.385693556 -0600
> @@ -1,9 +1,10 @@
>  /lib64/ld-linux-x86-64.so.2 
>  libc.so.6 => /lib64/libc.so.6 
>  libdl.so.2 => /lib64/libdl.so.2 
> +libEGL.so.1 => /lib64/libEGL.so.1 
>  libGLdispatch.so.0 => /lib64/libGLdispatch.so.0 
> -libGL.so.1 => /lib64/libGL.so.1 
> -libGLX.so.0 => /lib64/libGLX.so.0 
> +libGLESv2.so.2 => /lib64/libGLESv2.so.2 
> +libm.so.6 => /lib64/libm.so.6 
>  libX11.so.6 => /lib64/libX11.so.6 
>  libXau.so.6 => /lib64/libXau.so.6 
>  libxcb.so.1 => /lib64/libxcb.so.1 

Those are the only lib differences, I don't know if this is useful information at all, but there it is.  Sadly I can't dedicate a card to this for passthrough.

**Comment by ebassi** [comment]
Since qemu is using an EGL extension, looking at the output of `glxinfo` is pointless.

What you're seeing are the effects of a wrongly used EGL extension. The qemu code seems to be doing compile time detection of some DRM extension, instead of doing run time detection as it should.

I'm pretty sure that the nvidia binary blob driver does not support dmabuf in any way, which means that any attempt at using an EGL extension that passes around dmabuf objects is bound to fail.

The appropriate fix is for qemu is to use the appropriate API only if the extension that provides those symbols is available *at run time*. Epoxy has no other responsibility other than dlopen()-ing libGL and providing shims to its entry points.

I strongly encourage you to open an issue on the qemu bug tracking system instead of trying to "fix" Epoxy.

**Comment by nwnk** [comment]
> It appears that nvidia is not exporting this symbol via conventional means. es2_info reports the extension's availability, but glxinfo does not.

No, that is the conventional means. What that's telling you is that nvidia's driver supports that extension in EGL contexts but not in GLX contexts. Presumably virgl is asking for a GLX context?

**Comment by prodrigestivill** [comment]
> This is not an Epoxy issue, as Epoxy is working exactly as intended.

Since `virt-viewer` is using `spicy-gtk`, a new issue has been created in their issue tracker as [spice-gtk#100](https://gitlab.freedesktop.org/spice/spice-gtk/issues/100).
